### ISSUE TITLE
Export Format don't appear in Lizmap

### ISSUE THREAD

**Issue by Maverick83000** [3liz/qgis-wfsOutputExtension#15]
Hello,

I installed the HelloWorld et the wfsOutputExtension plugins and configured the fcgid conf file of apache2 ("_/etc/apache2/conf-available/serve-cgi-bin.conf_") like describes in the documentation.

When I try the url on the Hello World plugin (_http://151.236.35.125/cgi-bin/qgis_mapserv.fcgi?SERVICE=Hello_), it works and I've a response.
But when I try the export on lizmap, I just have GeoJson and GML formats ...

Qgis version : 3.4.13
Lizmap : 3.3.6
Lizmap plugin : 3.2.13
Qgis-server : 3.4.15

Thanks for your help.

**Comment by Gustry** [comment]
Can you try a request `SERVICE=ATLAS&REQUEST=GETCAPABILITIES` to check that the plugin is well installed?
Did you clear the cache of the getcapabilities response?

**Comment by Maverick83000** [comment]
Hi Gustry,

Thanks for your answer. 
The request with "SERVICE=ATLAS&REQUEST=GETCAPABILITIES" don't work:
_<ServiceExceptionReport version="1.3.0">
<ServiceException code="Service configuration error">Service unknown or unsupported</ServiceException>
</ServiceExceptionReport>_

But it works with "SERVICE=WMS&VERSION=1.3.0&REQUEST=GetCapabilities"

**Comment by Gustry** [comment]
> The request with "SERVICE=ATLAS&REQUEST=GETCAPABILITIES" don't work:

and 

> But it works with "SERVICE=WMS&VERSION=1.3.0&REQUEST=GetCapabilities"

They are two different requests. You must have the first one working. It means the plugin is not well installed.
Did you put the folder `wfsOutputExtension` next to the other one? (named like this)

**Comment by Maverick83000** [comment]
I put the extension folder in the _/opt/qgis-server/plugins_ just near the folder _qgis-helloserver_
![image](https://user-images.githubusercontent.com/11660738/86011728-60425800-ba1d-11ea-9b4e-efe0bf8ab17e.png)

**Comment by Maverick83000** [comment]
I said something wrong early, 
I modified the file _/etc/apache2/mods-available/fcgid.conf_ for adding the **QGIS_PLUGINPATH**
Maybe should I modified the file _/etc/apache2/conf-available/serve-cgi-bin.conf_ ?

**Comment by Glasnost46** [comment]
J'ai le même problème avec : 

QGIS 3.10.8
Lizmap 3.3.9
QGIS server 3.10.7.1
WFSoutputExtension 1.3.0 ou 1.2.3 (les deux versions ont été testées)
Windows server 2012

Le dossier du plugin est déposé ici : 
\OSGeo4W64\apps\qgis-ltr\python\plugins\ 

Le getCapabilities ne mentionne que les formats de base GeoJson et GML

Je suis preneur d'une piste parce que je n'en ai plus moi même :)

Merci

**Comment by rldhont** [comment]
Take a look at the QGIS Server Log, you can check if the plugin is installed and well loaded.

**Comment by Glasnost46** [comment]
Thank you for your answer.

Indeed, here is my log : 

QGIS version:
3.10.8-A Coruña 'A Coruña', d09a4f2388

Python path:
['C:\\OSGeo4W64\\apps\\qgis-ltr\\python', 'C:\\OSGeo4W64\\apps\\qgis-ltr\\python\\plugins', 'C:\\OSGeo4W64\\apps\\Python37\\DLLs', 'C:\\OSGeo4W64\\ap', 'C:\\OSGeo4W64\\bin\\python37.zip', 'C:\\OSGeo4W64\\apps\\Python37\\lib', 'C:\\OSGeo4W64\\apps\\qgis-ltr\\bin', 'C:\\OSGeo4W64\\apps\\Python37', 'C:\\OSGeo4W64\\apps\\Python37\\lib\\site-packages', 'C:\\OSGeo4W64\\apps\\Python37\\lib\\site-packages\\win32', 'C:\\OSGeo4W64\\apps\\Python37\\lib\\site-packages\\win32\\lib', 'C:\\OSGeo4W64\\apps\\Python37\\lib\\site-packages\\Pythonwin']
16:42:31 WARNING Python error[19460]: An error occurred during execution of following code:
qgis.utils.initServerInterface(465349856352)

Traceback (most recent call last):
  File "", line 1, in 
  File "C:/OSGeo4W64/apps/qgis-ltr/./python\qgis\utils.py", line 612, in initServerInterface
    from qgis.server import QgsServerInterface
ImportError: cannot import name 'QgsServerInterface' from 'qgis.server' (unknown location)

Python version:
3.7.0 (v3.7.0:1bf9cc5093, Jun 27 2018, 04:59:51) [MSC v.1914 64 bit (AMD64)]

QGIS version:
3.10.8-A Coruña 'A Coruña', d09a4f2388

Python path:
['C:/OSGeo4W64/apps/qgis-ltr/./python', 'C:/Windows/system32/config/systemprofile/AppData/Roaming/QGIS/QGIS3\\profiles\\default/python', 'C:\\OSGeo4W64\\apps\\qgis-ltr\\python\\plugins', 'C:/Windows/system32/config/systemprofile/AppData/Roaming/QGIS/QGIS3\\profiles\\default/python/plugins', 'C:/OSGeo4W64/apps/qgis-ltr/./python/plugins', 'C:\\OSGeo4W64\\apps\\qgis-ltr\\python', 'C:\\OSGeo4W64\\apps\\qgis-ltr\\python\\plugins', 'C:\\OSGeo4W64\\apps\\Python37\\DLLs', 'C:\\OSGeo4W64\\ap', 'C:\\OSGeo4W64\\bin\\python37.zip', 'C:\\OSGeo4W64\\apps\\Python37\\lib', 'C:\\OSGeo4W64\\apps\\qgis-ltr\\bin', 'C:\\OSGeo4W64\\apps\\Python37', 'C:\\OSGeo4W64\\apps\\Python37\\lib\\site-packages', 'C:\\OSGeo4W64\\apps\\Python37\\lib\\site-packages\\win32', 'C:\\OSGeo4W64\\apps\\Python37\\lib\\site-packages\\win32\\lib', 'C:\\OSGeo4W64\\apps\\Python37\\lib\\site-packages\\Pythonwin']
16:42:31 WARNING Python error[19460]: Traceback (most recent call last):
  File "C:/OSGeo4W64/apps/qgis-ltr/./python\qgis\utils.py", line 634, in startServerPlugin
    server_plugins[packageName] = package.serverClassFactory(serverIface)
  File "C:\OSGeo4W64\apps\qgis-ltr\python\plugins\wfsOutputExtension\__init__.py", line 35, in serverClassFactory
    return WfsOutputExtensionServer(serverIface)
  File "C:\OSGeo4W64\apps\qgis-ltr\python\plugins\wfsOutputExtension\wfs_output_extension_server.py", line 14, in __init__
    from .wfs_filter import WFSFilter
  File "C:/OSGeo4W64/apps/qgis-ltr/./python\qgis\utils.py", line 743, in _import
    mod = _builtin_import(name, globals, locals, fromlist, level)
  File "C:\OSGeo4W64\apps\qgis-ltr\python\plugins\wfsOutputExtension\wfs_filter.py", line 21, in <module>
    from qgis.server import QgsServerFilter
ImportError: cannot import name 'QgsServerFilter' from 'qgis.server' (unknown location)

16:42:31 WARNING Couldn't load server plugin wfsOutputExtension due to an error when calling its serverClassFactory() method[19460]: Couldn't load server plugin wfsOutputExtension due to an error when calling its serverClassFactory() method


ImportError: cannot import name 'QgsServerFilter' from 'qgis.server' (unknown location)


Traceback (most recent call last):
  File "C:/OSGeo4W64/apps/qgis-ltr/./python\qgis\utils.py", line 634, in startServerPlugin
    server_plugins[packageName] = package.serverClassFactory(serverIface)
  File "C:\OSGeo4W64\apps\qgis-ltr\python\plugins\wfsOutputExtension\__init__.py", line 35, in serverClassFactory
    return WfsOutputExtensionServer(serverIface)
  File "C:\OSGeo4W64\apps\qgis-ltr\python\plugins\wfsOutputExtension\wfs_output_extension_server.py", line 14, in __init__
    from .wfs_filter import WFSFilter
  File "C:/OSGeo4W64/apps/qgis-ltr/./python\qgis\utils.py", line 743, in _import
    mod = _builtin_import(name, globals, locals, fromlist, level)
  File "C:\OSGeo4W64\apps\qgis-ltr\python\plugins\wfsOutputExtension\wfs_filter.py", line 21, in 
    from qgis.server import QgsServerFilter
ImportError: cannot import name 'QgsServerFilter' from 'qgis.server' (unknown location)

Thank you

**Comment by Gustry** [comment]
@Glasnost46 Your problem is quite different.
QGIS Server detected the plugin and tried to load it.
But it fails on the Python API:

> ImportError: cannot import name 'QgsServerFilter' from 'qgis.server' (unknown location)

I have never tried to install a QGIS plugin server on Windows.

 Did you use https://github.com/NaturalGIS/qgis_server_and_lizmap_on_windows ?
CC @gioman, do you know about installing QGIS Server plugin on Windows?

**Comment by Glasnost46** [comment]
Thanks for your reply Gustry.
I have use the lizmap website to install all (https://docs.lizmap.com/current/fr/install/windows.html), everything seems to work fine except this plugin. My getcapabilities requests are ok.

**Comment by gioman** [comment]
>  CC @gioman, do you know about installing QGIS Server plugin on Windows?

@Gustry  I have not tested QGIS Server plugins on Windows, nevertheless it would intersting to do so on a machine where QGIS Server was installed and configured with the "recipe" https://github.com/NaturalGIS/qgis_server_and_lizmap_on_windows @Glasnost46 could you do it?

**Comment by Glasnost46** [comment]
> @Glasnost46 could you do it?

I think my setup is ok, except for this plugin.
This plugin worked on windows2012 in my previous setup with qgis2.8 / lizmap 3.2 / qgisserver

**Comment by gioman** [comment]
> qgisserver

QGIS Server 2 or 3?

**Comment by Glasnost46** [comment]
> 
> 
> > qgisserver
> 
> QGIS Server 2 or 3?

With QGIS server 2.18.28

**Comment by gioman** [comment]
QGIS Server 3 was basically rewritten from scratch, so there is a chance that something is not working with plugins on Windows.

**Comment by Gustry** [comment]
@Glasnost46 In your comment, you are talking about QGIS 3:
https://github.com/3liz/qgis-wfsOutputExtension/issues/15#issuecomment-668065825

At the top, we can read 

> QGIS version: 3.10.8-A Coruña 'A Coruña', d09a4f2388

Where did you get this from?

If you are really running QGIS Server 2 (so you must have QGIS Desktop 2 too !), the AtlasPrint plugin must be an older version maybe like 1.1.0 https://github.com/3liz/qgis-atlasprint/releases That's you got the error "ImportError: cannot import name 'QgsServerFilter' from 'qgis.server' (unknown location)"

**Comment by Glasnost46** [comment]
> 
> 
> @Glasnost46 In your comment, you are talking about QGIS 3:
> [#15 (comment)](https://github.com/3liz/qgis-wfsOutputExtension/issues/15#issuecomment-668065825)
> 
> At the top, we can read
> 
> > QGIS version: 3.10.8-A Coruña 'A Coruña', d09a4f2388
> Where did you get this from?

From my actual qgis_server.log
 
> If you are really running QGIS Server 2 (so you must have QGIS Desktop 2 too !), the AtlasPrint plugin must be an older version maybe like 1.1.0 https://github.com/3liz/qgis-atlasprint/releases That's you got the error "ImportError: cannot import name 'QgsServerFilter' from 'qgis.server' (unknown location)"

Now I'm using qgis server 3.10.7.1 with Qgis 3.10.8 and the plugin doesn't work.
Before, with qgis server 2.18.28 and qgis 2.8 it worked well.

I haven't the atlas plugin in my plugin folder.

Anyway, a conflict between old and new files is quite possible I guess.

Thanks for your help

**Comment by Maverick83000** [comment]
Hi.
I'm back from my Hollidays and the cruel life.
I will try to install the plugin again and the atlasprint plugin, but it failed to. 

I try to had those lines to the _000-default.conf_ but the qgis-server crashed ...

_<VirtualHost *:80>

 ServerAdmin webmaster@localhost
 DocumentRoot /var/www/html

 ErrorLog ${APACHE_LOG_DIR}/error.log
 CustomLog ${APACHE_LOG_DIR}/access.log combined

 FcgidInitialEnv DISPLAY ":99"
 FcgidInitialEnv QGIS_SERVER_LOG_LEVEL "0"
 FcgidInitialEnv QGIS_SERVER_LOG_STDERR "1"

 <Location /qgisserver>
  SetHandler fcgid-script
  FcgidWrapper /usr/lib/cgi-bin/qgis_mapserv.fcgi virtual
  Options +ExecCGI -MultiViews +FollowSymLinks
  Require all granted
 </Location>

</VirtualHost>_

**Comment by Gustry** [comment]
Did you make some progress?

>  I will try to install the plugin again and the atlasprint plugin, but it failed to.

Do you have any error message?

Check : 
* [ ] Where QGIS is looking for plugins?
* [ ] If the folder is correct, `wfsOutputExtension` with the `metadata.txt` inside. Better to take the zip from release
* [ ] Check the rights on the file. `root` might not be good when unzipping

**Comment by aeduard** [comment]
I've also tried to enable wfsOutputExtension qgis-server plug-in. At first I had difficulties, because it was not clear to me where should I put the QGIS_PLUGINPATH line (having Apache2 with FastCGI). I've checked qgis-server documentation [here](https://docs.qgis.org/3.10/en/docs/user_manual/working_with_ogc/server/plugins.html#http-server-configuration). I've modified my configuration at sites-available, than conf-available (serve-cgi-bin.conf). Nothing happened, GET CAPABILITIES not worked. It was not clear for me which file should I modify.

Than I started to check each qgis-server plug-in documentation I could find, and at [Atlasprint](https://github.com/3liz/qgis-atlasprint/blob/master/atlasprint/README.md) plug-in I've found clearly specified **_/etc/apache2/mods-available/fcgid.conf_**
There I've pasted the following lines, reload and restart apache2 service and GET CAPABILITIES and lizmap export started to work.

```
FcgidInitialEnv LC_ALL "en_US.UTF-8"
FcgidInitialEnv PYTHONIOENCODING UTF-8
FcgidInitialEnv LANG "en_US.UTF-8"
FcgidInitialEnv QGIS_SERVER_LOG_FILE /tmp/qgis-000.log
FcgidInitialEnv QGIS_SERVER_LOG_LEVEL 0
FcgidInitialEnv QGIS_PLUGINPATH "/opt/qgis-server/plugins"
```
I've extracted as root user, seems it's not an issue.
Still having problems with accented/special characters, as it doesn't get recognized at export. I've thought the first 3 lines in the above config would help, but it didn't. But generally the plug-in is working, and I'm happy with it.
